http_parser.c

```diff
@@ -472,6 +483,89 @@ static struct {
 
 int http_message_needs_eof(const http_parser *parser);
 
+const char* find_crlf(const char* p, const char* data, size_t len);
+
+#if defined(_MSC_VER) // SSE2 is baseline
+#include <intrin.h>
+#define __builtin_ctz _tzcnt_u32
+#elif defined(__SSE2__) && defined(__GNUC__)
+#include <emmintrin.h>
+#endif
+
+#ifndef USE_INTRISICS_CRLF
+# define USE_INTRISICS_CRLF defined(_MSC_VER) || (defined(__SSE2__) && defined(__GNUC__))
+#endif
+
+#if USE_INTRISICS_CRLF
+
+int32_t get_crlf_mask(const char* p) {
+  /* [ c, 0, 0, 0, 0, 0 .. 0 ] */
+  const __m128i vCR = _mm_set1_epi8(0x0d);
+
+  /* [ c, 0, 0, 0, 0, 0 .. 0 ] */
+  const __m128i vLF = _mm_set1_epi8(0x0a);
+
+  __m128i v1, v2;
+  v1 = *(__m128i*)p;
```

Comment:
Aliases.  It's also not directly obvious to me if it respects alignment.

Reply:
Hmm... any recommendations on how would you prefer this to be done for the alignment part? The alignment is ensured in find_crlf and it is done just once before the loop. Should I document it or write it in some comments or do you prefer if I move the alignment code in here? I guess that would incur some unnecessary alignment operations though

Reply:
(GCC defines the vector types with `__may_alias__`, making this safe. If you're worried about memory alignment you could use `_mm_lddqu_si128` explicitly.)

Reply:
I have changed the current version to looping unaligned bytes and anything less than 32 one by one, maybe you guys can also review the new commit instead? anyways _mm_lddqu_si128 is unfortunately sse3, right now i think we are just using sse2 for baseline

Reply:
@zbjornson I did try it out and loaded the sse3 headers, such a trivial change actually gave me a consistent 4% improvement. Unfortunately I think we might be rather out of luck though.

Reply:
Why out of luck? I don't see a problem with requiring SSE3, which was introduced in early 2004.

Reply:
@mscdex I am not sure what is the "baseline" for the supported architectures, I decided to try making use of sse2 instruction sets only because v8 requires sse2, so I suppose that was the baseline, if we are actually cool with supporting higher versions of vector instructions then great, I will change it immediately! :)

Reply:
I forgot that `lddqu` is SSE3. Sticking to SSE2 is nice because it's baseline on all platforms, and there's not much else you can do to get a perf bump here until you get to 256-bit registers (AVX) or get `pshufb` (SSSE3).

As far as improving node's performance specifically, node's build would have to be tweaked to emit for that instruction set. None of the oldest [currently supported x86/64 platforms](https://github.com/nodejs/node/blob/master/BUILDING.md#supported-platforms-1) require more than SSE2 AFAIK. It would be nice if node.js was open to requiring minimum CPU features (especially ones older than 8-10 years); there's a lot that could be done to improve performance if that were to happen.

I'm a bit surprised that `lddqu` is that much faster, that should only be the case if you're routinely loading across a CL, and I wouldn't expect this code to be memory-bound overall. (The benchmark I think always uses the same payload, so I guess not surprising.) I'll look in more detail today ... I'm also wondering if doing alignment is beneficial overall given that headers tend to be short.

Reply:
@zbjornson I was rather surprised too, such a simple operation. Actually with what I have modified, so we dont end up reading out of bounds, is basically the same as how memchr is implemented with sse2, but i guess its good enough.

I actually tried a avx2 version, with ymm registers and the performance improvements wasnt great, rather surprising. I tested it on a mid 2014 mbp 15" with haswell family i7, maybe things gets faster on the newer ones, not too sure.

Reply:
@zbjornson by the way I think headers can get rather large, think about cookies, i read somewhere typical http headers are between 200bytes to 2kb nowadays.

Reply:
As you have it now, I'm not seeing a difference between `lddqu` and `movdqa` (I don't think `lddqu` is "supposed" to be faster than `movdqa` in the case of CL splits with 16B-alignment).

As for AVX2, I get a 7-8% speedup over the current SSE2 version, and appears to be bottlenecking on branches. (`find_crlf` is down to <5% of the total instructions retired though. `parse_url_char` is the next heaviest hitter, which is also unsurprisingly bottlenecked on branches. There's maybe some stuff to be done with LUTs or `pshufb` in that function, for another time.)

Reply:
@zbjornson yup the branches is a huge bottleneck, the function is just so large.

Some of the stuff I've added that shaved off about 5%(not the SIMD part) was actually reverse engineered from clang's compiled version, and then converted back to c, gcc(i didnt look at the asm from the latest gcc though, was using gcc6 i think) seems to be rather weak when it comes to optimizing certain branches.

But definitely did a much better job at optimizing the large function as a whole compared to clang(cant remember the version either)

I actually wrote one more version, just to test it out using PCMPESTRI, it was significantly slower, god knows why :P